File: SampleApp/SampleApp-objc/backendCommunication/TokenHandler.m
#import "TokenHandler.h"

@interface TokenHandler ()

@end

@implementation TokenHandler

@synthesize clientSecret = _clientSecret;
@synthesize clientId = _clientId;

static NSString *baseUrl = @"https://eu.customer.365id.com/api/v1/";
const NSInteger minimumSecondsToUseToken = 25;
static NSString *storedTokenDataKey = @"TokenData365ID";

- (id)initWithData
    :(NSString*)accessToken
    :(NSString*)expiresIn
    :(NSString*)refreshToken
    :(NSString*)expireTimeTxt
    :(NSString*)clientSecret
    :(NSString*)clientId
{
    self = [super init];
    self.accessToken = accessToken;
    self.expiresIn = expiresIn;
    self.refreshToken = refreshToken;
    self.expireTimeTxt = expireTimeTxt;
    self.clientSecret = clientSecret;
    self.clientId = clientId;
    return self;
}

- (NSString*)getAccessToken {
    return self.accessToken;
}

- (NSString*)getRefreshToken {
    return self.refreshToken;
}

- (void)setExpireTimeTxt {
    NSInteger useSeconds;
    if (self.expiresIn == nil) {
        useSeconds = 0;
    }
    else {
        NSInteger expiresInInteger = [self.expiresIn integerValue];
        useSeconds = expiresInInteger;
    }
    NSDate *expireTime = [NSCalendar.currentCalendar dateByAddingUnit:NSCalendarUnitSecond value:useSeconds toDate:[[NSDate alloc] init] options:0];
    if (expireTime == nil) {
        return;
    }
    NSDateFormatter *dateFormatter = [self getDateFormatter];
    NSString *result = [dateFormatter stringFromDate:expireTime];
    self.expireTimeTxt = result;
}

- (void)setClientSecret:(NSString *)value {
    _clientSecret = value;
}

- (void)setClientId:(NSString *)value {
    _clientId = value;
}

- (NSString*)getClientSecret {
    return _clientSecret;
}

- (NSString*)getClientId {
    return _clientId;
}

/** Get a UTC date formatter */
- (NSDateFormatter*)getDateFormatter {
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    NSTimeZone *timeZone = [NSTimeZone timeZoneWithName:@"UTC"];
    if (timeZone != nil) {
        [dateFormatter setTimeZone:timeZone];
    }
    [dateFormatter setDateFormat:@"yyyy-MM-dd HH:mm:ss"];
    return dateFormatter;
}

/** Get seconds left of active token */
- (NSInteger)getTokenSecondsLeft {
    NSString *dateTxt = self.expireTimeTxt;
    if (dateTxt == nil) {
        return 0;
    }
    NSDateFormatter *dateFormatter = [self getDateFormatter];
    NSDate *futureDate = [dateFormatter dateFromString:dateTxt];
    if (futureDate == nil) {
        return 0;
    }
    NSDateComponents *dateComponents = [NSCalendar.currentCalendar components:NSCalendarUnitSecond fromDate:[[NSDate alloc] init] toDate:futureDate options:0];
    NSInteger secondsLeft;
    if (dateComponents != nil) {
        secondsLeft = dateComponents.second;
    }
    else {
        secondsLeft = 0;
    }
    if (secondsLeft < 0) {
        secondsLeft = 0;
    }
    return secondsLeft;
}

/** Check if the token is valid based on the expire date */
- (BOOL)isValidToken {
    NSInteger secondsLeft = [self getTokenSecondsLeft];
    BOOL result = secondsLeft >= minimumSecondsToUseToken;
    return result;
}

/** Create TokenData object from json */
+ (TokenHandler*)jsonToTokenData:(NSString *)json {
    NSData* jsonData = [json dataUsingEncoding:NSUTF8StringEncoding];
    if (jsonData == nil) {
        return nil;
    }
    NSError *error;
    NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:jsonData options:NSJSONReadingMutableContainers error:&error];
    if (error != nil) {
        return nil;
    }
    NSString *accessToken = [dict objectForKey:@"access_token"];
    NSString *expiresInTxt = [dict objectForKey:@"expires_in"];
    NSString *refreshToken = [dict objectForKey:@"refresh_token"];
    NSString *expireTimeTxt = [dict objectForKey:@"expireTimeTxt"];
    NSString *clientSecret = [dict objectForKey:@"clientSecret"];
    NSString *clientId = [dict objectForKey:@"clientId"];
    TokenHandler *tokenHandler = [[TokenHandler alloc] initWithData:accessToken :expiresInTxt :refreshToken :expireTimeTxt :clientSecret :clientId];
    return tokenHandler;
}

/** Create json from TokenData object */
+ (NSString*)tokenDataToJson:(TokenHandler *)tokenData {
    NSDictionary *details = @{
        @"access_token":tokenData.accessToken,
        @"expires_in":tokenData.expiresIn,
        @"refresh_token":tokenData.refreshToken,
        @"expireTimeTxt":tokenData.expireTimeTxt,
        @"clientSecret":tokenData.clientSecret,
        @"clientId":tokenData.clientId,
    };
    NSError *parseError;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:details options:NSJSONWritingPrettyPrinted error:&parseError];
    if (parseError != nil) {
        return nil;
    }
    NSString *json = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    return json;
}

/** Save the TokenData on the device */
+ (BOOL)save
    :(NSString*)clientSecret
    :(NSString*)clientId
    :(NSString*)responseBody
{
    TokenHandler *tokenData = [self jsonToTokenData:responseBody];
    if (tokenData == nil) {
        return false;
    }
    [tokenData setExpireTimeTxt];
    [tokenData setClientId:clientId];
    [tokenData setClientSecret:clientSecret];
    NSString *updatedJson = [self tokenDataToJson:tokenData];
    if (updatedJson == nil) {
        return false;
    }
    [NSUserDefaults.standardUserDefaults setValue:updatedJson forKey:storedTokenDataKey];
    return true;
}

/** Load the TokenData from the device */
+ (TokenHandler*)loadData {
    NSString *json = [NSUserDefaults.standardUserDefaults stringForKey:storedTokenDataKey];
    if (json == nil) {
        return nil;
    }
    TokenHandler *tokenData = [self jsonToTokenData:json];
    if (tokenData == nil) {
        [self delete];
        return nil;
    }
    return tokenData;
}

/** Delete the TokenData from the device */
+ (void)delete{
    [NSUserDefaults.standardUserDefaults removeObjectForKey:storedTokenDataKey];
}

/** Delete TokenData if Credentials has been changed */
+ (void)deleteIfCredentialsChanged
    :(NSString*)clientSecret
    :(NSString*)clientId
{
    TokenHandler *tokenData = [self loadData];
    if (tokenData == nil) {
        return;
    }
    NSString *storedClientId = [tokenData getClientId];
    NSString *storedClientSecret = [tokenData getClientSecret];
    if (![storedClientId isEqualToString:clientId] || ![storedClientSecret isEqualToString:clientSecret]) {
        [self delete];
    }
}

/** Get a token for starting the SDK */
+ (void)getToken
    :(NSString *)clientSecret
    :(NSString *)clientId
    :(completionHandlerType)completionHandler
{
    [self deleteIfCredentialsChanged:clientSecret :clientId];
    TokenHandler *storedTokenData = [self loadData];

    if (storedTokenData != nil) {
        NSInteger secondsLeft = [storedTokenData getTokenSecondsLeft];
        NSString *secondsLeftTxt = [NSString stringWithFormat: @"%ld", (long)secondsLeft];
        NSLog(@"getToken() - %@ seconds left of token.", secondsLeftTxt);
        if ([storedTokenData isValidToken]) {
            NSLog(@"getToken() - Using stored token.");
            completionHandler([storedTokenData getAccessToken]);
            return;
        }
    }

    NSString *action;
    NSDictionary *bodyParams;
    NSString *bearer;
    if (storedTokenData != nil) {
        action = @"refresh_token";
        bodyParams = @{
            @"refresh_token":[storedTokenData getRefreshToken]
        };
        bearer = [storedTokenData getAccessToken];
        [self delete];
    }
    else {
        action = @"access_token";
        bodyParams = @{
            @"client_id":clientId,
            @"client_secret":clientSecret
        };
        bearer = nil;
    }

    NSString *fullUrl = [NSString stringWithFormat:@"%@%@", baseUrl, action];
    NSURL *url = [[NSURL alloc] initWithString:fullUrl];
    if (url == nil) {
        completionHandler(nil);
        return;
    }
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:url];
    request.timeoutInterval = 5;
    request.HTTPMethod = @"POST";
    request.cachePolicy = NSURLRequestReloadIgnoringLocalCacheData;
    [request addValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    if (bearer != nil) {
        NSString *bearerTxt = [NSString stringWithFormat:@"Bearer %@", bearer];
        [request setValue:bearerTxt forHTTPHeaderField:@"Authorization"];
    }
    NSError *parseParamsError;
    NSData *jsonParams = [NSJSONSerialization dataWithJSONObject:bodyParams options:NSJSONWritingPrettyPrinted error:&parseParamsError];
    if (parseParamsError != nil) {
        completionHandler(nil);
        return;
    }
    request.HTTPBody = jsonParams;
    [self runRequest:request :clientSecret :clientId :completionHandler];
}

/** Get a token for starting the SDK */
+ (void)runRequest
    :(NSMutableURLRequest *) request
    :(NSString *)clientSecret
    :(NSString *)clientId
    :(completionHandlerType)completionHandler
{
    [[[NSURLSession sharedSession] dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error)
    {
        NSHTTPURLResponse *httpStatus = (NSHTTPURLResponse *)response;
        if (error != nil || httpStatus.statusCode != 200) {
            NSLog(@"getToken() - error:");
            NSLog(@"%@", error.localizedDescription);
            NSLog(@"%@", httpStatus.description);
            dispatch_async(dispatch_get_main_queue(), ^{
                completionHandler(nil);
            });
            return;
        }

        NSError *parseError;
        NSDictionary *responseDictionary = [NSJSONSerialization JSONObjectWithData:data options:0 error:&parseError];
        if (parseError != nil || responseDictionary == nil) {
            NSLog(@"getToken() - error:");
            NSLog(@"%@", parseError.description);
            dispatch_async(dispatch_get_main_queue(), ^{
                completionHandler(nil);
            });
            return;
        }

        NSError *parseDataError;
        NSData *jsonData = [NSJSONSerialization dataWithJSONObject:responseDictionary options:NSJSONWritingPrettyPrinted error:&parseDataError];
        if (parseDataError != nil || jsonData == nil) {
            NSLog(@"getToken() - error:");
            NSLog(@"%@", parseDataError.description);
            dispatch_async(dispatch_get_main_queue(), ^{
                completionHandler(nil);
            });
            return;
        }

        NSString *responseBody = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
        if (responseBody == nil) {
            NSLog(@"getToken() - error: Could not store response json.");
            dispatch_async(dispatch_get_main_queue(), ^{
                completionHandler(nil);
            });
            return;
        }

        if (![self save:clientSecret :clientId :responseBody]) {
            NSLog(@"getToken() - error: Could not save response.");
            dispatch_async(dispatch_get_main_queue(), ^{
                completionHandler(nil);
            });
            return;
        }

        NSLog(@"getToken() - success!");
        dispatch_async(dispatch_get_main_queue(), ^{
            TokenHandler *newData = [self loadData];
            if (newData == nil) {
                completionHandler(nil);
            }
            else {
                completionHandler([newData getAccessToken]);
            }
        });
    }] resume];
}
@end

File: SampleApp/SampleApp/BackendCommunication/TokenHandler.swift
import UIKit

// MARK: - Handle Token requests
class TokenHandler {
    
    private static let baseUrl: String = "https://eu.customer.365id.com/api/v1/"
    
    // MARK: The token object
    private struct TokenData : Codable {
        
        private let accessToken: String?
        private let expiresIn: Int?
        private let refreshToken: String?
        private var expireTimeTxt: String?
        private var clientSecret: String?
        private var clientId: String?
        private let minimumSecondsToUseToken: Int = 25
        private static let storedTokenDataKey: String = "TokenData365ID"
        
        public func getAccessToken() -> String? {
            self.accessToken
        }
        
        public func getRefreshToken() -> String {
            self.refreshToken ?? ""
        }
        
        private mutating func setExpireTimeTxt() {
            guard let expireTime = Calendar.current.date(byAdding: .second, value: self.expiresIn ?? 0, to: Date()) else { return }
            self.expireTimeTxt = getDateFormatter().string(from: expireTime)
        }
        
        private mutating func setClientSecret(_ value: String) {
            self.clientSecret = value
        }
        
        private mutating func setClientId(_ value: String) {
            self.clientId = value
        }
        
        private func getClientSecret() -> String? {
            self.clientSecret
        }
        
        private func getClientId() -> String? {
            self.clientId
        }
        
        /** Keys to handle from the response */
        private enum CodingKeys: String, CodingKey {
            case accessToken = "access_token"
            case expiresIn = "expires_in"
            case refreshToken = "refresh_token"
            case expireTimeTxt = "expireTimeTxt"
            case clientSecret = "clientSecret"
            case clientId = "clientId"
        }
        
        /** Parse json to objects */
        public init(from decoder: Decoder) throws {
            let values = try decoder.container(keyedBy: CodingKeys.self)
            accessToken = try values.decodeIfPresent(String.self, forKey: .accessToken)
            expiresIn = try values.decodeIfPresent(Int.self, forKey: .expiresIn)
            refreshToken = try values.decodeIfPresent(String.self, forKey: .refreshToken)
            expireTimeTxt = try values.decodeIfPresent(String.self, forKey: .expireTimeTxt)
            clientSecret = try values.decodeIfPresent(String.self, forKey: .clientSecret)
            clientId = try values.decodeIfPresent(String.self, forKey: .clientId)
        }
        
        /** Get a UTC date formatter */
        private func getDateFormatter() -> DateFormatter {
            let dateFormatter = DateFormatter()
            if let timeZone = TimeZone(identifier: "UTC") {
                dateFormatter.timeZone = timeZone
            }
            dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
            return dateFormatter
        }
        
        /** Get seconds left of active token */
        public func getTokenSecondsLeft() -> Int {
            guard let dateTxt = self.expireTimeTxt else { return 0 }
            guard let futureDate = getDateFormatter().date(from: dateTxt) else { return 0 }
            var secondsLeft = Calendar.current.dateComponents([.second], from: Date(), to: futureDate).second ?? 0
            if secondsLeft < 0 {
                secondsLeft = 0
            }
            return secondsLeft
        }
        
        /** Check if the token is valid based on the expire date */
        public func isValidToken() -> Bool {
            getTokenSecondsLeft() >= self.minimumSecondsToUseToken
        }
        
        /** Create TokenData object from json */
        private static func jsonToTokenData(_ json: String) -> TokenData? {
            guard let data: Data = json.data(using: .utf8) else { return nil }
            return try? JSONDecoder().decode(TokenData.self, from: data)
        }
        
        /** Create json from TokenData object */
        private static func tokenDataToJson(_ tokenData: TokenData) -> String? {
            guard let jsonData = try? JSONEncoder().encode(tokenData) else { return nil }
            return String(data: jsonData, encoding: .utf8)
        }
        
        /** Save the TokenData on the device */
        public static func save(clientSecret: String, clientId: String, responseBody: String) -> Bool {
            guard var tokenData = jsonToTokenData(responseBody) else { return false }
            tokenData.setExpireTimeTxt()
            tokenData.setClientId(clientId)
            tokenData.setClientSecret(clientSecret)
            guard let updatedJson = tokenDataToJson(tokenData) else { return false }
            UserDefaults.standard.set(updatedJson, forKey: storedTokenDataKey)
            return true
        }
        
        /** Load the TokenData from the device */
        public static func load() -> TokenData? {
            guard let json = UserDefaults.standard.string(forKey: storedTokenDataKey) else { return nil }
            guard let tokenData = jsonToTokenData(json) else {
                delete()
                return nil
            }
            return tokenData
        }
        
        /** Delete the TokenData from the device */
        public static func delete() {
            UserDefaults.standard.removeObject(forKey: storedTokenDataKey)
        }
        
        /** Delete TokenData if Credentials has been changed */
        public static func deleteIfCredentialsChanged(clientSecret: String, clientId: String) {
            guard let tokenData = TokenData.load() else { return }
            let storedClientId = tokenData.getClientId()
            let storedClientSecret = tokenData.getClientSecret()
            if storedClientId != clientId || storedClientSecret != clientSecret {
                TokenData.delete()
            }
        }
    }
    
    /** Get a token for starting the SDK */
    public static func getToken(clientSecret: String, clientId: String, _ completionHandler: @escaping (_ token: String?) -> Void) {
        
        TokenData.deleteIfCredentialsChanged(clientSecret: clientSecret, clientId: clientId)
        let storedTokenData = TokenData.load()
        
        if let tokenData = storedTokenData {
            print("getToken() - \(tokenData.getTokenSecondsLeft()) seconds left of token.")
            if tokenData.isValidToken() {
                print("getToken() - Using stored token.")
                completionHandler(tokenData.getAccessToken())
                return
            }
        }
        
        let action: String
        let bodyParams: [String: Any]
        let bearer: String?
        if let tokenData = storedTokenData {
            action = "refresh_token"
            bodyParams = [
                "refresh_token": tokenData.getRefreshToken()
            ]
            bearer = tokenData.getAccessToken()
            TokenData.delete()
        }
        else{
            action = "access_token"
            bodyParams = [
                "client_id": clientId,
                "client_secret": clientSecret
            ]
            bearer = nil
        }
        
        guard let url = URL(string: "\(self.baseUrl)\(action)") else {
            completionHandler(nil)
            return
        }
        var request = URLRequest(url: url)
        request.timeoutInterval = 5
        request.httpMethod = "POST"
        request.cachePolicy = .reloadIgnoringLocalCacheData
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        if let bearerData = bearer {
            request.setValue("Bearer \(bearerData)", forHTTPHeaderField: "Authorization")
        }
        guard let jsonParams = try? JSONSerialization.data(withJSONObject: bodyParams) else {
            completionHandler(nil)
            return
        }
        request.httpBody = jsonParams
        
        let mainThread = DispatchQueue.main
        let currentThread = Thread.current
        URLSession.shared.dataTask(with: request) { [weak mainThread, weak currentThread] (data, response, error) in
            guard mainThread != nil && currentThread != nil else { return }
            let httpStatus = response as? HTTPURLResponse
            guard
                error == nil,
                httpStatus?.statusCode == 200,
                let receivedData = data,
                let responseBody = String(data: receivedData, encoding: .utf8),
                !responseBody.isEmpty,
                TokenData.save(clientSecret: clientSecret, clientId: clientId, responseBody: responseBody)
            else {
                print("getToken() - error:")
                print("\(error?.localizedDescription ?? "N/A")")
                print("\(httpStatus?.description ?? "N/A")")
                mainThread?.async {
                    completionHandler(nil)
                }
                return
            }
            print("getToken() - success!")
            mainThread?.async {
                completionHandler(TokenData.load()?.getAccessToken() ?? nil)
            }
        }.resume()
    }
}
